book-2-platform-api/chapters/INTEGRATION_TESTS.md

```diff
@@ -0,0 +1,279 @@
+# Automated Testing
+
+## Overview
+
+How do you know your code does what you think it does?
+
+_You Test It!_
+
+There are two basic approaches to testing.
+
+### Manual Testing
+
+You or _**someone else**_ uses the application and verify that it does what it's supposed to do. Usually a QA person will have a giant checklist of fetures to test and they will click, scroll and type into your app to make sure each feature works.
+
+### Automated Testing
+
+You _**write code**_ that runs your application and verifies that it does what it's supposed to do.
+
+The trend in the industry is toward automated testing.
+
+## Types of Automated Tests
+
+There are two main types of automated testing.
+1. Unit Testing - Testing individual classes or methods.
+1. Integration Testing - Testing a complete request/response cycle. An integration test will use and literally visit a URL in the web application. It will fill out HTML forms, click buttons and follow links to make things happen in your application. Data will be read from and written to the database.
+
+We will be covering integration tests in this chapter.
+
+## Steps for each Automated Test
+
+* Arrange - Building out the classes and methods needed to make meet the assertion.
+* Act - Each test should have the code required to test an action, or feature, of your program.
+* Assert - What should be true, or what state should an object be in when code is executed.
+
+
+## Integration Testing in ASP.NET Core MVC
+
+It's true that a human _could_ open a web browser and manually perform all the steps necessary to test all parts of a web application, but - like with many things - a computer can do it faster, with fewer mistakes, and without complaining about how boring it is to test and retest the same app all the time.
+
+
+https://docs.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-2.1
+
+
+## Example
+
+Think waaaaaay back to when you learned React and consider the process of creating a new `Animal` resource in your `json-server` API.
+
+First you needed to build a JSON representation of an animal that would be boarded in a kennel. Here's a basic version.
+
+```js
+const newAnimal = {
+    "name": "Jack",
+    "breed": "Cocker Spaniel",
+    "age": 4,
+    "hasShots": true
+}
+```
+
+Then you had to perform an HTTP operation of type `POST` in order to save that resource in your API.
+
+```js
+fetch("http://localhost:3000/animals", {
+    "method": "POST",
+    "headers": {
+        "Content-Type": "application/json"
+    },
+    "body": JSON.stringify(newAnimal)
+})
+```
+
+## Process
+
+What's happening during this process?
+
+1. An HTTP client sends a `POST` request to the `/animal` route.
+1. The `Post()` method of the `AnimalController` is called.
+1. ADO.NET is used to insert the data into the corresponding Animal table in the database.
```

Comment:
It may seem confusing, but making the connection back to something that they have already used quite a bit and making that the anchor for learning something new is a useful mechanism - especially for concrete learners.  New knowledge given in the context of previously acquired knowledge is what they need.